Fix #5486 & part of #5343: Introducing new wiki page for code coverage usage and limitations (#5483)
oppia/oppia-android: wiki/Oppia-Android-Code-Coverage.md

> ## Table of Contents

Reviewer: I think it's important to mention somewhere that the team's philosophy (and implementation) of code coverage is that we only analyze a source unit within the context of its own tests (that is, we don't care if a utility has incidental code coverage because another, unrelated test references it). This actually makes code coverage percentage numbers much more valuable since they force developers to ensure a unit itself is directly and well covered via its own tests.

Author: Got it. Added this section to address this: https://github.com/oppia/oppia-android/blob/e97b1abe555b08c633c5300f01b7ea7dd1e1f359/wiki/Oppia-Android-Code-Coverage.md#unit-centric-coverage-philosophy, but not sure where to place it. Initially considered placing it right at the start in the 'Understanding Code Coverage' section, but also felt that including it there might confuse or complicate things for those starting out. So placed it as a separate section. Would adjust the contents and position as per feedback.

Reviewer: I think the placement that you chose seems reasonable--I don't have any thoughts on it needing to be anywhere else. Thanks for adding it!

        "Positive" 
    } else { 
        "Negative"
    }
}
```

A test is created to verify that this function works correctly:
Please refer to the [Oppia Android testing documentation](https://github.com/oppia/oppia-android/wiki/Oppia-Android-Testing) to learn more about writing tests.

```
import org.junit.Test
import kotlin.test.assertEquals

class checkSignTest {
    @Test
    fun testCheckSign_withPositiveInteger_returnsPositive() {
        assertEquals("Positive", checkSign(4))
    }
}
```

This test checks whether passing a positive integer '+4' `checkSign(4)` correctly returns "Positive" as output.
However, it doesn’t test how the function handles negative numbers. This means that only a portion of the function’s behavior is being tested, leaving the handling of negative numbers unverified.

For thorough testing, the tests should also check how the function responds to negative inputs, this can be done by adding a test case by passing a negative integer to expect a return of "Negative" as output.

```diff
import org.junit.Test
import kotlin.test.assertEquals

class checkSignTest {
    @Test
    fun testCheckSign_withPositiveInteger_returnsPositive() {

>         assertEquals("Positive", checkSign(4))

Reviewer: Here & above: we should use Truth assertThat rather than assertEquals (to avoid being inconsistent with tests seen throughout the codebsae). This is different than the test cases where you use assertEquals in generated code since this is an actual reference that may be copied during development.

Author: Ah Got it. Replaced `assertEquals` with `assertThat`

Coverage Analysis: **FAIL** :x: <br>

#

- **Number of Files Assessed:** <br>This displays the number of Kotlin files included in the coverage analysis. Files with other extensions, such as .xml, .json, or .md, are excluded from this analysis.

- **Overall Coverage:** <br>This reflects the average code coverage percentage for the files modified in the current pull request. It provides a snapshot of how thoroughly the new code is tested.

- **Coverage Analysis:** <br>This indicates whether the pull request passes the coverage analysis. A check mark denotes a passing result, while an X mark indicates a failure.

### 2. Report Details Section

#

**2.a. Failure Cases**

### Failure Cases

| File | Failure Reason | Status |
|------|----------------|:------:|
| File.kt | No appropriate test file found for File.kt. | :x: |

#

This section lists files that failed to acquire coverage data. Failures may occur due to various reasons, including:

- The absence of an appropriate test file for the respective source file, preventing coverage analysis.
- Bazel failing to collect coverage data for the source file.
- Bazel lacking a reference to the required source file in the collected data.

> - Other potential reasons are still under exploration.

Reviewer: Need a newline after this, I think, to avoid the next line being included in this bullet item.

Author: Done. Added a new line.

>   The table has two columns:

Reviewer: Technically three, right? Since there's a status column.

Author: Yeah, sorry. Was stuck with the old template. 

Updated it to have three columns:

3. **Status:** Indicates that the coverage status is a failure.

1. **File:** Displays the file for which the error occurred (Clicking the drop-down reveals the file's path in the codebase).
2. **Failure Reason:** Describes the reason for the failure.

> Note: If this table or section is not present in the coverage report, it indicates that no changes exhibited failure

Reviewer: ```suggestion
Note: If this table or section is not present in the coverage report, it indicates that no changes exhibited failure.
```

Author: Done.

#

**2.b. Failing coverages**

### Failing coverage

| File | Coverage | Lines Hit | Status | Min Required |
|------|:--------:|----------:|:------:|:------------:|
| <details><summary>Fail1.kt</summary>scripts/src/java/org/oppia/android/scripts/coverage/Fail1.kt</details> | 51.38% | 148 / 288 | :x: | 70% |
| <details><summary>Fail2.kt</summary>utility/src/main/java/org/oppia/android/util/Fail2.kt</details> | 77.78% | 7 / 9 | :x: | 80% _*_ |
| <details><summary>Fail3.kt</summary>domain/src/main/java/org/oppia/android/domain/classify/Fail3.kt</details> | 10.00% | 1 / 10 | :x: | 30% _*_ |

>**_*_** represents tests with custom overridden pass/fail coverage thresholds

#

> This section highlights files that have failed to meet the minimum coverage percentage. Any file that does not meet the minimum threshold percentage is considered to have a failing status. The minimum threshold is set to a standard value (currently as 70%) and displayed alongside the status for each file.

Reviewer: We ought to link to where the threshold is defined rather than fixing it to 70% here (that way the documentation doesn't need to be changed when the threshold is updated in the future).

Author: Yeah, so can I link to the entire CoverageReporter.kt file (since that is where it is staying now), I believe we don't want to link to specific line of MIN_THRESHOLD as that could change?

Updating it to link to the entire CoverageReporter.kt file (will update as per feedback).

Reviewer: I think the link you added makes sense and looks good.

The data presented includes:

1. **File:** Name of the file that met the coverage threshold (Clicking the drop-down reveals the file's path in the codebase).
2. **Coverage Percentage:** Percentage of coverage achieved for the file.
3. **Lines Covered:** Number of lines covered by the test cases, shown in "lines hit / lines found" format.
4. **Status:** Indicates whether the coverage status is a pass or fail.
5. **Required Percentage:** Minimum percentage required to pass the coverage check for the file.

**2.d. Exempted coverages**

### Exempted coverage
<details open><summary>Files exempted from coverage</summary> <br>

| File | Failure Reason |
|------|----------------|
| <details><summary>TestExemptedFile.kt</summary>app/src/main/java/org/oppia/android/app/activity/TestExemptedFile.kt</details> | This file is exempted from having a test file; skipping coverage check. |
| <details><summary>SourceIncompatible.kt</summary>app/src/main/java/org/oppia/android/app/activity/SourceIncompatible.kt</details> | This file is incompatible with code coverage tooling; skipping coverage check. |

</details>

#

Certain files are exempt from coverage checks. These exemptions include:

> 1. **Test File Exemptions:** Files that do not have corresponding test files are exempt from coverage checks. Since no test files are available for these sources, coverage analysis cannot be performed, and these files are therefore skipped.

Reviewer: ```suggestion
1. **Test File Exemptions:** Files are exempted from having corresponding test files are also exempt from coverage checks. Since no test files are available for these sources, coverage analysis cannot be performed, and these files are therefore skipped.
```

Clarification since there's another interpretation for the previous wording: that tests are simply absent (which is actually a failure when not exempted, as you discuss in the page above).

Author: Yeah, updated to be more specific to exempted cases. Thanks

2. **Source File Incompatibility Exemptions:** Some files are currently incompatible with Bazel coverage execution ([see tracking issue #5481](https://github.com/oppia/oppia-android/issues/5481)) and are temporarily excluded from coverage checks.

You can find the complete list of exemptions in this file: [test_file_exemptions.textproto](https://github.com/oppia/oppia-android/blob/develop/scripts/assets/test_file_exemptions.textproto)

> This section appears at the bottom of the report, as a drop-down. It includes a table listing the exemptions as,

Reviewer: ```suggestion
This section appears at the bottom of the report, as a drop-down. It includes a table listing the exemptions with columns:
```

For consistency with other table explanations on this page.

Author: Done. Updated as suggested.

1. **File:** Displays the file name that is exempted (Clicking the drop-down reveals the file's path in the codebase).
2. **Failure Reason:** Describes the specific reason for its exemption.

With this report, you can review the coverage percentages for various files and identify which files pass or fail the coverage check.

## 2. Local Command-Line Interface (CLI) Tools

While the CI check provides an overview of coverage, you might want to visualize how tests cover specific files locally, including which lines are covered and which are not. For this, Oppia's local command-line coverage tooling is useful.

> Note: Follow these [Bazel setup instructions](https://github.com/oppia/oppia-android/wiki/Oppia-Bazel-Setup-Instructions) if Bazel isn't yet set up on your device:

Reviewer: ```suggestion
Note: Follow these [Bazel setup instructions](https://github.com/oppia/oppia-android/wiki/Oppia-Bazel-Setup-Instructions) if Bazel isn't yet set up in your local development environment.
```

'Device' usually refers to mobile phones in the context of Android development, so this is hopefully a bit clearer.

Author: Ah, understood. Updated as suggested.

Oppia Android allows you to generate coverage reports in HTML format using the command:

### Run Coverage

```
bazel run //scripts:run_coverage -- <path_to_root> <list_of_relative_path_to_files>
```

- <path_to_root>: Your root directory.
- <list_of_relative_path_to_files>: Files you want to generate coverage reports for.

For example, to analyze coverage for the file MathTokenizer.kt, use the relative path:

```
bazel run //scripts:run_coverage -- $(pwd) utility/src/main/java/org/oppia/android/util/math/MathTokenizer.kt
```

By default, this will generate an HTML report in the coverage_reports directory. For the given file, the report will be saved as **coverage_reports/utility/src/main/java/org/oppia/android/util/math/MathTokenizer/coverage.html**

> A list of files can be provided as an input to generate coverage reports for all the files.

Reviewer: ```suggestion
A list of files can be provided as an input to generate coverage reports for each of the provided the files. An example of this is:
```

Some small changes for clarity.

Author: Thanks, updated as suggested.

```
bazel run //scripts:run_coverage -- $(pwd) 
utility/src/main/java/org/oppia/android/util/math/MathTokenizer.kt 
utility/src/main/java/org/oppia/android/util/parser/math/MathModel.kt
```

### Process Timeout

The default wait time for a process to complete is 300 seconds (5 minutes). If the process does not finish within this period, you may encounter a "Process Did Not Finish Within Intended Time" error. To extend the wait time, you can modify the timeout setting by adding the flag --processTimeout=15 (The Time unit is minutes).

> ```

Reviewer: Here & elsewhere for CLI examples, you can use 'sh' for shell syntax highlighting (it doesn't add much, but it makes it a bit more explicit to the viewer that is's a shell command).

Author: Sure. Updated it for every bazel run command. 
Also, is it necessary to do the same for kt files too, ie. to add ```kotlin or can I leave it in the default way.

Reviewer: I think it'd be nice for all code blocks, and I actually left a newer comment for Kotlin (so resolving this one in favor of that one).

bazel run //scripts:run_coverage -- $(pwd)
utility/src/main/java/org/oppia/android/util/parser/math/MathModel.kt --processTimeout=15
```

## 2.1 Understanding the CI Coverage Report

> Let's examine the coverage.html report located at coverage_reports/utility/src/main/java/org/oppia/android/util/math/MathTokenizer/coverage.html to understand its contents.

Reviewer: ```suggestion
Let's examine the example coverage.html report (generated using the sample command in the previous section) to understand its contents. The report should be located at ``coverage_reports/utility/src/main/java/org/oppia/android/util/math/MathTokenizer/coverage.html``.
```

Some rewording to try and improve clarity.

Author: Thanks, updated as suggested.

**Header Section:** At the top of the report, you'll find information including:

![image](https://github.com/user-attachments/assets/08c4f0a0-51b5-4d8e-a2f3-ff13d8b6cb2f)

> #

Reviewer: Did you mean for these (here & below) to be headers? I think this section needs some formatting work since it looks a bit awkward right now in Markdown mode.

Author: By Header, I was trying to reference the top summary section. I think I cluttered the section.
But still not sure how to name the 2 portions, for now referenced them as:
- Coverage Overview
- Line-by-Line Coverage Breakdown

and tried reformatting it, does it look less cluttered or need further work?

Reviewer: I think it looks good; I'll add additional suggestions for reorganizing if I think of any.

- **File Path:** The relative path of the file to the oppia-android codebase.
- **Coverage Percentage:** The percentage of code covered by tests.
- **Line Coverage Data:** Details on the coverage of individual lines.
- **Main Content:**

> Visual Coverage Representation: The report visually highlights which lines are covered and which are not.

Reviewer: ```suggestion
Visual Coverage Representation: The report visually highlights the coverage status of source code lines.
```

Attempt at improving clarity.

Author: Thanks, much clearer to me as well. Updated with the same.

![image](https://github.com/user-attachments/assets/0e36fde3-639b-4874-b809-59d33827388d)

#

- **Red Lines:** Indicate lines not covered by test cases.
- **Green Lines:** Represent lines that are covered by test cases.

These generated html reports are be used to identify areas of your code that may need additional testing.

## Increasing Code Coverage Metrics

> While figuring out which lines are covered and which are not, you can increase code coverage by ensuring that each line is thoroughly tested. To enhance code coverage effectiveness, locate the test files associated with the source file (typically found in the same package but within the test directories).

Reviewer: Locate the files, and then do what? I think it's missing something like "analyze and improve test coverage" with a link to the other page that you're introducing in this PR.

Author: Oops, was thinking of following this line with the "By identifying and adding appropriate test scenarios to cover previously uncovered lines, you will boost the coverage percentage and improve the overall quality and reliability of your code." line.

But now I have now rephrased the paragraph and moved the 'Note' section to the end.

> Note: Some files in the app module may have multiple test files, located in the sharedTest and test packages, all testing a single source file. (eg: The file StateFragment.kt has 2 test files StateFragmentTest.kt under sharedTest and StateFragmentLocalTest.kt under test folder)

Reviewer: ```suggestion
Note: Some files in the app module may have multiple test files, located in the sharedTest and test packages, all testing a single source file. For example: StateFragment.kt has 2 test files StateFragmentTest.kt under ``app/src/sharedTest`` and StateFragmentLocalTest.kt under ``app/src/test``.
```

Some formatting and wording alteration for clarity.

Also, please add permalinks to each of the three referenced files.

Author: Thanks, updated as suggested and added links for all 3 files.

By identifying and adding appropriate test scenarios to cover previously uncovered lines, you will boost the coverage percentage and improve the overall quality and reliability of your code.

## Limitations of the code coverage tool

> 1. **Incompatibility with Code Coverage Analysis:** Certain test targets in the Oppia-Android codebase fail to execute and collect coverage using the Bazel coverage command. The underlying issues are still being investigated ([see tracking issue #5481](https://github.com/oppia/oppia-android/issues/5481)), and these files are currently exempt from coverage checks.

Reviewer: ```suggestion
1. **Incompatibility with Code Coverage Analysis:** Certain test targets in the Oppia-Android codebase fail to execute and collect coverage using the Bazel coverage command. The underlying issues are still being investigated ([see tracking issue #5481](https://github.com/oppia/oppia-android/issues/5481)), and these files are currently exempt from coverage checks. However, it's expected that all new test files should work without needing this exemption.
```

Clarification to help set precedent.

Author: Got it. Updated as suggested.

- **Line Coverage Data:** Details on the coverage of individual lines.
- **Main Content:**

Visual Coverage Representation: The report visually highlights which lines are covered and which are not.

![image](https://github.com/user-attachments/assets/0e36fde3-639b-4874-b809-59d33827388d)

#

- **Red Lines:** Indicate lines not covered by test cases.
- **Green Lines:** Represent lines that are covered by test cases.

These generated html reports are be used to identify areas of your code that may need additional testing.

## Increasing Code Coverage Metrics

While figuring out which lines are covered and which are not, you can increase code coverage by ensuring that each line is thoroughly tested. To enhance code coverage effectiveness, locate the test files associated with the source file (typically found in the same package but within the test directories).

Note: Some files in the app module may have multiple test files, located in the sharedTest and test packages, all testing a single source file. (eg: The file StateFragment.kt has 2 test files StateFragmentTest.kt under sharedTest and StateFragmentLocalTest.kt under test folder)

> By identifying and adding appropriate test scenarios to cover previously uncovered lines, you will boost the coverage percentage and improve the overall quality and reliability of your code.

Author: This section on approval will add a link to the writing effective tests or writing tests with good behavioural coverage page.

Reviewer: I think two paragraphs above might be a good place for the link (see my comment). Maybe also move this sentence to that paragraph or drop it since the section will probably be a bit easier to read with the 'note' being the last item.

Author: Done, rephrased the above paragraph to include this too thereby positioning the `Note` section to the end.

Reviewer: Great--do you think this can be resolved now or is further action needed?

Author: Thanks, resolving it with the link added to the referenced wiki - https://github.com/oppia/oppia-android/pull/5483/files#diff-696bbeaab338ccd0f2816eba3863990fe6fad20ec44ef5bec8475950eea26b76R377
